src/backend/distributed/planner/distributed_planner.c

```diff
@@ -193,75 +240,39 @@ distributed_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
 	ReplaceTableVisibleFunction((Node *) parse);
 
 	/* create a restriction context and put it at the end if context list */
-	PlannerRestrictionContext *plannerRestrictionContext =
-		CreateAndPushPlannerRestrictionContext();
+	ctx.plannerRestrictionContext = CreateAndPushPlannerRestrictionContext();
+
+	/*
+	 * We keep track of how many times we've recursed into the planner, primarily
+	 * to detect whether we are in a function call. We need to make sure that the
+	 * PlannerLevel is decremented exactly once at the end of the next PG_TRY
+	 * block, both in the happy case and when an error occurs.
+	 */
+	PlannerLevel++;
+
 
 	PG_TRY();
 	{
-		/*
-		 * We keep track of how many times we've recursed into the planner, primarily
-		 * to detect whether we are in a function call. We need to make sure that the
-		 * PlannerLevel is decremented exactly once at the end of this PG_TRY block,
-		 * both in the happy case and when an error occurs.
-		 */
-		PlannerLevel++;
-
-		/*
-		 * For trivial queries, we're skipping the standard_planner() in
-		 * order to eliminate its overhead.
-		 *
-		 * Otherwise, call into standard planner. This is required because the Citus
-		 * planner relies on both the restriction information per table and parse tree
-		 * transformations made by postgres' planner.
-		 */
+		DistributedPlan *delegatedPlan = NULL;
+		bool hasExternParam = false;
 
-		if (needsDistributedPlanning && FastPathRouterQuery(originalQuery))
+		if (fastPathRouterQuery)
```

Comment:
To make the flow a little more explicit, what if we do something like:

```C
	if (fastPathRouterQuery)
	{
		result = PlanFastPathDistributedStmt(&ctx, distributionKeyValue);
	}
	else
	{
		PlannedStmt *localPlan = standard_planner(ctx->parse, ctx->cursorOptions,
											   ctx->boundParams);

		else if (needsDistributedPlanning)
		{
			result = PlanDistributedStmt(&ctx, localPlan, rangeTableList, rteIdCounter);
		}
		else if ((delegatedPlan = TryToDelegateFunctionCall(ctx.parse, 
															&hasExternParam)))
		{
			result = PlanDelegatedFunctionCall(&ctx, localPlan, delegatedPlan, hasExternParam);
		}
		else
		{
			result = localPlan;
		}
	}
``` 

I don't have strong preference, but still wanted to share if you find it easier to follow as well, maybe do it

Reply:
changed it to something close to this now, because I store the localPlan in the planContext.